File: src/pontil_ffi.mjs
import {
  ExitCode$isFailure,
  ExitCode$isSuccess,
  Linux,
  MacOS,
  Other,
  Windows,
} from "./pontil/types.mjs";

export function setExitCode(value) {
  if (ExitCode$isFailure(value)) {
    setExitCode_(1);
  } else if (ExitCode$isSuccess(value)) {
    setExitCode_(0);
  } else if (typeof value == "number") {
    setExitCode_(value);
  }
}

function setExitCode_(value) {
  if (globalThis?.Deno) {
    globalThis.Deno.exitCode = value;
  } else if (globalThis?.process) {
    globalThis.process.exitCode = value;
  }
}

export function stop() {}

export function isWindows() {
  return rawPlatform() === "win32";
}

export function isMacos() {
  return rawPlatform() === "darwin";
}

export function isLinux() {
  return rawPlatform() === "linux";
}

export function osType() {
  const p = rawPlatform();
  switch (p) {
    case "win32":
      return new Windows();
    case "darwin":
      return new MacOS();
    case "linux":
      return new Linux();
    default:
      return new Other(p);
  }
}

export function osArch() {
  if (globalThis?.Deno) {
    return globalThis.Deno.build.arch;
  }
  return globalThis?.process?.arch ?? "unknown";
}

function rawPlatform() {
  if (globalThis?.Deno) {
    const os = globalThis.Deno.build.os;
    return os === "windows" ? "win32" : os;
  }
  return globalThis?.process?.platform ?? "linux";
}

File: src/pontil_ffi.erl
-module(pontil_ffi).
-export([
    is_linux/0,
    is_macos/0,
    is_windows/0,
    os_arch/0,
    os_type/0,
    set_exit_code/1,
    stop/0
]).

set_exit_code(ExitCode) ->
    Code =
        case ExitCode of
            failure -> 1;
            success -> 0
        end,
    erlang:put(pontil_process_exit_code, Code),
    nil.

stop() ->
    Code =
        case erlang:get(pontil_process_exit_code) of
            undefined -> 0;
            V -> V
        end,
    erlang:halt(Code).

is_windows() ->
    case os:type() of
        {win32, _} -> true;
        _ -> false
    end.

is_macos() ->
    case os:type() of
        {unix, darwin} -> true;
        _ -> false
    end.

is_linux() ->
    case os:type() of
        {unix, linux} -> true;
        _ -> false
    end.

os_type() ->
    case os:type() of
        {win32, _} -> windows;
        {unix, darwin} -> mac_o_s;
        {unix, linux} -> linux;
        {unix, Other} -> {other, atom_to_binary(Other, utf8)};
        _ -> {other, <<"unknown">>}
    end.

os_arch() ->
    list_to_binary(erlang:system_info(system_architecture)).
